### PR TITLE
Use MEF to load all plugins

### PR DESCRIPTION
**Pull request by sharwell**
MEF was previously used only for a subset of the types used in GitExtensions extensibility. This change extends the current implementation to also cover `IGitPlugin`. It does not make any changes to the way MEF is used (e.g. by switching to vs-mef).

The goal of this change is to consolidate functionality so it can be uniformly converted to a higher-performance framework in a follow-up commit.

~~🚧 I observed a possible behavior discrepancy in the top-level menu when testing this. I need to do more testing before I'm confident this is correct.~~ (Fixed in 14b331a293361e954ffc97ed4f68b132dfc52ba6)

📝 If accepted, please do not rebase or squash this pull request during the merge.

### CODE REVIEW

**Comment by vbjay** [comment]
It was mentioned that we could use vs-mef.  I think evaluating which one is
used would be a good idea.  See
https://github.com/Microsoft/vs-mef/blob/master/doc/mef_library_differences.md

**Comment by vbjay** [comment]
Another thought is do we want shadow copying so plugins aren't locked.  User modifies a plugin file and the app is running scenario.  https://www.codeproject.com/Articles/633140/MEF-and-AppDomain-Remove-Assemblies-On-The-Fly, an article that describes that scenario.  I don't know if it would work with vs-mef, if we went that route.

**Comment by gerhardol** [comment]
FYI: Tried this quickly in the debugger: If I add breakpoint in LoadPlugins() it works OK.
Otherwise either GitHub menu is not visible and occasionaly plugins are not loaded (nothing in Plugin menu or Settings-Plugins).

**Comment by sharwell** [comment]
@gerhardol That's consistent with my observation. I'll see about figuring out the timing issue after work. Since I had this in a branch I figured I'd get it posted as a starting point.

File: GitUI/CommandsDialogs/FormBrowse.cs [hunk 1]
```diff
@@ -459,25 +459,6 @@ private void _indexWatcher_Changed(object sender, IndexChangedEventArgs e)
             });
         }
 
-        private bool _pluginsLoaded;
-        private void LoadPluginsInPluginMenu()
-        {
-            if (_pluginsLoaded)
```
Comment (sharwell) [inline GitUI/CommandsDialogs/FormBrowse.cs]: @gerhardol The bad state we observed was caused by this condition. It should have been a condition similar to the following:

```
if (_pluginsLoaded || !PluginLoader.HasLoadedPlugins)
{
    return;
}
```

I ended up changing the "populate on demand" to a "populate when ready" path.

**Comment by jbialobr** [comment]
The GitHub plugin menu items are not loaded into the main menu.
It is related to
https://github.com/gitextensions/gitextensions/blob/81df6fa1e9fe961f96abd7f863f4c66ac534c9a8/GitUI/CommandsDialogs/FormBrowse.cs#L620

**Comment by sharwell** [comment]
@jbialobr I don't understand your comment. Is there a change/bug you are hoping I can look at?

**Comment by jbialobr** [comment]
Before
gitext05
After
gitext
It happens only on fresh GE start. Once you hit Refresh button the GitHub menu is added.

**Comment by sharwell** [comment]
@jbialobr Fixed in befcb0c

**Comment by drewnoakes** [comment]
I haven't used MEF before but the code here looks as clean as I'd expect from @sharwell.

I did some performance testing around plugin loading in #4694 and thought I could run the same tests here. I know there's been discussion about MEF caching, but I still wanted to get a baseline.

I set up the same test as in #4694: `Program.Main` calls `LoadPlugins.Load` directly, and I wrap the call with a `System.Diagnostics.Stopwatch`.

This current MEF implementation takes over 5 seconds to load all plugins. The highest duration I saw was 16 seconds but that was an extreme outlier.

The previous code took closer to 800ms, and we could get it as low as 500ms I expect. For whatever reason the timings I got before were less volatile.

This PR is no longer filtering DLLs by file name. If I put that clause into `ManagedExtensibility.GetCompositionContainers` then the time drops to around 2,300 ms.

**Comment by RussKie** [comment]
I do not really have an opinion due to my lack of experience with MEF.
If there is no significant (i.e. visible) degradation of the app's load time - then let's rebase and merge it in.

**Comment by sharwell** [comment]
> If there is no significant (i.e. visible) degradation of the app's load time - then let's rebase and merge it in.

This change will make the application load time a bit worse (call it ~5 seconds difference), but it makes it easier to transition to vs-mef which will more than make up the difference. Let me know if the two-step approach is acceptable.

**Comment by RussKie** [comment]
Yes, I think it is a sensible approach.

**Comment by drewnoakes** [comment]
I would prefer to know that vs-mef delivers a perf increase.

This work stemmed from a discussion reducing plugin scanning from 800ms to around 500ms. Now it's up to 5,000ms (rough numbers, from my laptop).

With this PR we'd only be using MEF in a small portion of the app. It's not bringing a far-reaching benefit and has an impact on startup perf. If vs-mef brings this down considerably then that's fantastic, but shouldn't we have some numbers first? It's fine to take one step back, so long as we actually take two steps forwards. I'd vote for doing this all in one PR so we know where we'd end up.

**Comment by vbjay** [comment]
One thing that might speed up things is using the [DirectoryCatalog](https://msdn.microsoft.com/en-us/library/system.componentmodel.composition.hosting.directorycatalog(v=vs.110).aspx).

**Comment by vbjay** [comment]
Along with that.  We could implement the idea of requiring the plugins to have a known prefix or suffix in the filename.

gitextensions-{plugin name here}.dll
or 
{plugin name here}GEPlugin.dll 
or 
... 

Then adjust the directorycatalog to use that prefix/suffix.  

Another thought is inheriting from the catalog and organizing the plugins folder so that each plugin is in its own folder.  Plugins\CreateLocalBranches\{plugin files here} for example.  If there are any common dependencies, we can have a common dependencies folder.

**Comment by vbjay** [comment]
Example mef code

```c#
/Create an assembly catalog of the assemblies with exports
    var catalog = new AggregateCatalog(
        new AssemblyCatalog(Assembly.GetExecutingAssembly()),
        new AssemblyCatalog(Assembly.Load("My.Other.Assembly")),
        new DirectoryCatalog(path, "*.dll"));

    //Create a composition container
    _container = new CompositionContainer(catalog);  
```

**Comment by sharwell** [comment]
@drewnoakes @vbjay Assuming no filtering of filenames at all (or special naming, etc.), vs-mef drops the load time from ~1.85 seconds to ~0.4 seconds. However, there are enough nuances to the loader code that I'd still prefer to take it as a separate change.

I was measuring around `PluginLoader.Load()` for this test.

**Comment by vbjay** [comment]
Sure.  Didn't see [this](https://github.com/Microsoft/vs-mef/blob/master/doc/hosting.md#migrating-from-net-mef-or-nuget-mef).  If we use vs-mef, there is no directorycatalog.

**Comment by drewnoakes** [comment]
> I was measuring around PluginLoader.Load() for this test.

That's where I was measuring too.

Can't compare 0.4s on your machine with my numbers, but it does seem sensible, with room to improve via file name filtering.

If you want to do this as two stages and somehow a subsequent PR doesn't deliver the perf we want then we can always revert this.

**Pull request merged**